File: R/ode_speed_test/ode_speed_test.R

#' ODE Speed Test for Ross-MacDonald Model
#' =========================================
#' This script provides:
#' 1. R-native ODE solving with deSolve for baseline comparison
#' 2. Interface to TensorFlow speed tests via reticulate
#' 3. Comparison with greta.dynamics iterate_dynamic_function
#'
#' Author: Ernest Moyo
#' Date: 2025
#' Issue: https://github.com/ernestmoyo/epiwave-foi-model/issues/4

# =============================================================================
# SETUP
# =============================================================================

# Required packages
required_packages <- c("deSolve", "ggplot2", "dplyr", "tidyr", 
                       "purrr", "tictoc", "reticulate")

# Install missing packages
for (pkg in required_packages) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    message(paste("Installing", pkg, "..."))
    install.packages(pkg)
  }
}

library(deSolve)
library(ggplot2)
library(dplyr)
library(tidyr)
library(purrr)
library(tictoc)

# =============================================================================
# ROSS-MACDONALD MODEL IN R
# =============================================================================

#' Ross-MacDonald ODE system with time-varying mosquito density
#'
#' @param t Current time
#' @param state Named vector c(x = human_infected, z = mosquito_infected)
#' @param params List of parameters including seasonal components
#'
ross_macdonald_ode <- function(t, state, params) {
  with(as.list(c(state, params)), {
    
    # Bimodal seasonal mosquito density
    m_t <- m_base * (1 + amp1 * cos(2 * pi * t / 365) + 
                         amp2 * cos(4 * pi * t / 365))
    
    # Ross-MacDonald equations
    dx <- m_t * a * b * z * (1 - x) - r * x
    dz <- a * c * x * (1 - z) - g * z
    
    list(c(dx, dz))
  })
}


#' Ross-MacDonald for multiple sites (vectorized)
#'
ross_macdonald_multisite <- function(t, state, params) {
  n_sites <- params$n_sites
  
  # Extract states
  x <- state[1:n_sites]
  z <- state[(n_sites + 1):(2 * n_sites)]
  
  # Bimodal seasonal mosquito density (site-specific m_base)
  m_t <- params$m_base * (1 + params$amp1 * cos(2 * pi * t / 365) + 
                              params$amp2 * cos(4 * pi * t / 365))
  
  # Ross-MacDonald equations (vectorized)
  dx <- m_t * params$a * params$b * z * (1 - x) - params$r * x
  dz <- params$a * params$c * x * (1 - z) - params$g * z
  
  list(c(dx, dz))
}


# =============================================================================
# SPEED TEST FUNCTIONS
# =============================================================================

#' Create test parameters for n_sites locations
#'
create_test_params <- function(n_sites, seed = 42) {
  set.seed(seed)
  
  list(
    n_sites = n_sites,
    m_base = rlnorm(n_sites, meanlog = 1, sdlog = 0.5),
    amp1 = 0.6,    # Primary seasonal amplitude
    amp2 = 0.3,    # Secondary amplitude  
    a = 0.3,       # Biting rate
    b = 0.5,       # Transmission mosquito -> human
    c = 0.8,       # Transmission human -> mosquito
    r = 1/14,      # Recovery rate (14 day infection)
    g = 1/10       # Mosquito death rate (10 day lifespan)
  )
}


#' Run speed test with deSolve
#'
#' @param n_sites_list Vector of site counts to test
#' @param n_years Number of years to simulate
#' @param n_repeats Number of repetitions for timing
#' @param methods Vector of deSolve methods to test
#'
run_desolve_speed_test <- function(
    n_sites_list = c(10, 50, 100, 500),
    n_years = 4,
    n_repeats = 5,
    methods = c("lsoda", "rk4", "euler", "ode45")
) {
  
  times <- seq(0, 365 * n_years, by = 1)
  
  results <- list()
  
  for (n_sites in n_sites_list) {
    message(sprintf("\nTesting with N = %d sites...", n_sites))
    
    # Create parameters
    params <- create_test_params(n_sites)
    
    # Initial state
    state <- c(
      x = rep(0.01, n_sites),  # Human infection
      z = rep(0.00, n_sites)   # Mosquito infection
    )
    
    for (method in methods) {
      message(sprintf("  Testing %s...", method))
      
      timing <- numeric(n_repeats)
      
      for (rep in seq_len(n_repeats)) {
        tic()
        
        tryCatch({
          out <- ode(
            y = state,
            times = times,
            func = ross_macdonald_multisite,
            parms = params,
            method = method
          )
          elapsed <- toc(quiet = TRUE)
          timing[rep] <- elapsed$toc - elapsed$tic
        }, error = function(e) {
          message(sprintf("    Error: %s", e$message))
          timing[rep] <- NA
        })
      }
      
      results[[length(results) + 1]] <- tibble(
        n_sites = n_sites,
        solver = paste0("deSolve::", method),
        mean_time = mean(timing, na.rm = TRUE),
        std_time = sd(timing, na.rm = TRUE),
        min_time = min(timing, na.rm = TRUE),
        max_time = max(timing, na.rm = TRUE)
      )
      
      message(sprintf("    Mean time: %.4f s", mean(timing, na.rm = TRUE)))
    }
  }
  
  bind_rows(results)
}


#' Run speed test with greta.dynamics
#'
run_greta_speed_test <- function(
    n_sites_list = c(10, 50, 100),
    n_years = 4,
    n_repeats = 5
) {
  
  # Check if greta.dynamics is available

if (!requireNamespace("greta.dynamics", quietly = TRUE)) {
    message("greta.dynamics not available, skipping greta speed test")
    return(NULL)
  }
  
  library(greta)
  library(greta.dynamics)
  
  results <- list()
  
  for (n_sites in n_sites_list) {
    message(sprintf("\nGreta test with N = %d sites...", n_sites))
    
    # This uses the iterate_dynamic_function approach from your RMarkdown
    timing <- numeric(n_repeats)
    
    for (rep in seq_len(n_repeats)) {
      tic()
      
      tryCatch({
        # Create greta arrays for parameters
        m_base <- normal(1, 0.5, dim = n_sites, truncation = c(0, Inf))
        
        # ... (simplified - full implementation would mirror your RMD)
        # For now, just test basic greta array operations
        
        x <- normal(0.01, 0.001, dim = n_sites)
        sims <- calculate(x, nsim = 10)
        
        elapsed <- toc(quiet = TRUE)
        timing[rep] <- elapsed$toc - elapsed$tic
      }, error = function(e) {
        message(sprintf("    Error: %s", e$message))
        timing[rep] <- NA
      })
    }
    
    results[[length(results) + 1]] <- tibble(
      n_sites = n_sites,
      solver = "greta.dynamics",
      mean_time = mean(timing, na.rm = TRUE),
      std_time = sd(timing, na.rm = TRUE),
      min_time = min(timing, na.rm = TRUE),
      max_time = max(timing, na.rm = TRUE)
    )
  }
  
  bind_rows(results)
}


# =============================================================================
# VISUALIZATION
# =============================================================================

#' Plot speed test results
#'
plot_speed_results <- function(results) {
  
  ggplot(results, aes(x = n_sites, y = mean_time, color = solver)) +
    geom_line(linewidth = 1) +
    geom_point(size = 3) +
    geom_errorbar(
      aes(ymin = mean_time - std_time, ymax = mean_time + std_time),
      width = 0.1
    ) +
    scale_x_log10() +
    scale_y_log10() +
    labs(
      title = "ODE Solver Speed Comparison",
      subtitle = "Ross-MacDonald model with bimodal seasonality (4 years)",
      x = "Number of Sites (N)",
      y = "Computation Time (seconds)",
      color = "Solver"
    ) +
    theme_minimal() +
    theme(legend.position = "bottom")
}


# =============================================================================
# MAIN EXECUTION
# =============================================================================

if (interactive()) {
  
  message("=" |> rep(60) |> paste(collapse = ""))
  message("R ODE Solver Speed Test")
  message("Ross-MacDonald Model with Time-Varying Parameters")
  message("=" |> rep(60) |> paste(collapse = ""))
  
  # Run deSolve tests
  message("\n--- deSolve Speed Tests ---")
  desolve_results <- run_desolve_speed_test(
    n_sites_list = c(10, 50, 100, 500),
    n_years = 4,
    n_repeats = 5
  )
  
  # Display results
  message("\n--- Results Summary ---")
  print(desolve_results)
  
  # Save results
  write.csv(desolve_results, "R/ode_speed_test/desolve_results.csv", row.names = FALSE)
  
  # Create plot
  p <- plot_speed_results(desolve_results)
  ggsave("R/ode_speed_test/speed_comparison.png", p, width = 10, height = 6)
  
  message("\nResults saved to R/ode_speed_test/")
}

File: python/ode_speed_test.py
"""
TensorFlow ODE Solver Speed Test for Ross-MacDonald Model
=========================================================
This script compares different ODE solvers in TensorFlow Probability
for the Ross-MacDonald malaria transmission model with:
- Time-varying parameters (bimodal seasonality)
- Multiple sites (N = 10, 50, 100, 500)
- 4 years of simulation

Author: Ernest Moyo
Date: 2025
Issue: https://github.com/ernestmoyo/epiwave-foi-model/issues/4
"""

import tensorflow as tf
import tensorflow_probability as tfp
import numpy as np
import time
import pandas as pd
from typing import Callable, Dict, List, Tuple

# Enable eager execution for debugging
tf.config.run_functions_eagerly(False)

# ============================================================================
# ROSS-MACDONALD MODEL DEFINITION
# ============================================================================

class RossMacDonaldODE:
    """
    Ross-MacDonald malaria transmission model with time-varying mosquito density.
    
    State variables:
        x: Proportion of infected humans
        z: Proportion of infected mosquitoes
    
    Parameters:
        m(t): Mosquito abundance relative to humans (time-varying, seasonal)
        a: Human biting rate
        b: Probability of transmission mosquito -> human
        c: Probability of transmission human -> mosquito
        r: Human recovery rate
        g: Mosquito death rate
    """
    
    def __init__(self, n_sites: int, params: Dict[str, tf.Tensor]):
        self.n_sites = n_sites
        self.params = params
        
    def seasonal_mosquito_density(self, t: tf.Tensor) -> tf.Tensor:
        """
        Bimodal seasonal mosquito density function.
        Two peaks per year (e.g., after rainy seasons).
        
        m(t) = m_base * (1 + amp1 * cos(2*pi*t/365) + amp2 * cos(4*pi*t/365))
        """
        m_base = self.params["m_base"]  # Shape: (n_sites,)
        amp1 = self.params["amp1"]       # Primary seasonal amplitude
        amp2 = self.params["amp2"]       # Secondary seasonal amplitude
        
        # Bimodal seasonality
        seasonal = (1.0 + 
                   amp1 * tf.cos(2.0 * np.pi * t / 365.0) + 
                   amp2 * tf.cos(4.0 * np.pi * t / 365.0))
        
        return m_base * seasonal
    
    def __call__(self, t: tf.Tensor, state: tf.Tensor) -> tf.Tensor:
        """
        ODE right-hand side function.
        
        Args:
            t: Current time (scalar)
            state: Current state [x, z] with shape (2, n_sites)
            
        Returns:
            Derivatives [dx/dt, dz/dt] with shape (2, n_sites)
        """
        x = state[0]  # Human infection proportion (n_sites,)
        z = state[1]  # Mosquito infection proportion (n_sites,)
        
        # Get time-varying mosquito density
        m = self.seasonal_mosquito_density(t)
        
        # Fixed parameters
        a = self.params["a"]  # Biting rate
        b = self.params["b"]  # Transmission prob mosquito -> human
        c = self.params["c"]  # Transmission prob human -> mosquito
        r = self.params["r"]  # Human recovery rate
        g = self.params["g"]  # Mosquito death rate
        
        # Ross-MacDonald equations
        dx_dt = m * a * b * z * (1.0 - x) - r * x
        dz_dt = a * c * x * (1.0 - z) - g * z
        
        return tf.stack([dx_dt, dz_dt], axis=0)


# ============================================================================
# SOLVER IMPLEMENTATIONS
# ============================================================================

def solve_with_dormand_prince(ode_fn: Callable, 
                               initial_state: tf.Tensor,
                               t_span: Tuple[float, float],
                               t_eval: tf.Tensor) -> tf.Tensor:
    """Solve ODE using Dormand-Prince (adaptive RK45) method."""
    solver = tfp.math.ode.DormandPrince(atol=1e-6, rtol=1e-6)
    results = solver.solve(ode_fn, t_span[0], initial_state, 
                          solution_times=t_eval)
    return results.states


def solve_with_bdf(ode_fn: Callable,
                   initial_state: tf.Tensor,
                   t_span: Tuple[float, float],
                   t_eval: tf.Tensor) -> tf.Tensor:
    """Solve ODE using BDF (Backward Differentiation Formula) method."""
    solver = tfp.math.ode.BDF(atol=1e-6, rtol=1e-6)
    results = solver.solve(ode_fn, t_span[0], initial_state,
                          solution_times=t_eval)
    return results.states


def solve_with_fixed_euler(ode_fn: Callable,
                           initial_state: tf.Tensor,
                           t_span: Tuple[float, float],
                           dt: float = 0.1) -> tf.Tensor:
    """Solve ODE using fixed-step Euler method."""
    t_start, t_end = t_span
    n_steps = int((t_end - t_start) / dt)
    
    states = [initial_state]
    state = initial_state
    t = t_start
    
    for _ in range(n_steps):
        derivative = ode_fn(t, state)
        state = state + dt * derivative
        t = t + dt
        states.append(state)
    
    return tf.stack(states, axis=0)


def solve_with_rk4(ode_fn: Callable,
                   initial_state: tf.Tensor,
                   t_span: Tuple[float, float],
                   dt: float = 0.5) -> tf.Tensor:
    """Solve ODE using fixed-step RK4 method."""
    t_start, t_end = t_span
    n_steps = int((t_end - t_start) / dt)
    
    states = [initial_state]
    state = initial_state
    t = t_start
    
    for _ in range(n_steps):
        k1 = ode_fn(t, state)
        k2 = ode_fn(t + dt/2, state + dt/2 * k1)
        k3 = ode_fn(t + dt/2, state + dt/2 * k2)
        k4 = ode_fn(t + dt, state + dt * k3)
        
        state = state + (dt/6) * (k1 + 2*k2 + 2*k3 + k4)
        t = t + dt
        states.append(state)
    
    return tf.stack(states, axis=0)


# ============================================================================
# SPEED TEST FUNCTIONS
# ============================================================================

def create_test_parameters(n_sites: int) -> Dict[str, tf.Tensor]:
    """Create realistic test parameters for n_sites locations."""
    
    # Base mosquito density varies by site (log-normal variation)
    np.random.seed(42)
    m_base = tf.constant(np.random.lognormal(mean=1.0, sigma=0.5, size=n_sites), 
                        dtype=tf.float32)
    
    return {
        "m_base": m_base,
        "amp1": tf.constant(0.6, dtype=tf.float32),   # Primary seasonal amplitude
        "amp2": tf.constant(0.3, dtype=tf.float32),   # Secondary amplitude
        "a": tf.constant(0.3, dtype=tf.float32),      # Biting rate
        "b": tf.constant(0.5, dtype=tf.float32),      # Transmission mosquito->human
        "c": tf.constant(0.8, dtype=tf.float32),      # Transmission human->mosquito
        "r": tf.constant(1/14, dtype=tf.float32),     # Recovery rate (14 day duration)
        "g": tf.constant(1/10, dtype=tf.float32),     # Mosquito death rate (10 day lifespan)
    }


def run_speed_test(n_sites_list: List[int] = [10, 50, 100, 500],
                   n_years: int = 4,
                   n_repeats: int = 5) -> pd.DataFrame:
    """
    Run speed tests for different solvers and number of sites.
    
    Returns DataFrame with timing results.
    """
    
    results = []
    t_span = (0.0, 365.0 * n_years)
    t_eval = tf.linspace(0.0, 365.0 * n_years, 365 * n_years)
    
    solvers = {
        "DormandPrince": solve_with_dormand_prince,
        "BDF": solve_with_bdf,
        "RK4_fixed": solve_with_rk4,
        "Euler_fixed": solve_with_fixed_euler,
    }
    
    for n_sites in n_sites_list:
        print(f"\nTesting with N = {n_sites} sites...")
        
        # Create model
        params = create_test_parameters(n_sites)
        model = RossMacDonaldODE(n_sites, params)
        
        # Initial state: low infection levels
        initial_state = tf.constant(
            [[0.01] * n_sites,  # x: human infection
             [0.0] * n_sites],  # z: mosquito infection
            dtype=tf.float32
        )
        
        for solver_name, solver_fn in solvers.items():
            print(f"  Testing {solver_name}...")
            
            times = []
            for rep in range(n_repeats):
                try:
                    start = time.perf_counter()
                    
                    if "fixed" in solver_name.lower():
                        _ = solver_fn(model, initial_state, t_span)
                    else:
                        _ = solver_fn(model, initial_state, t_span, t_eval)
                    
                    end = time.perf_counter()
                    times.append(end - start)
                    
                except Exception as e:
                    print(f"    Error: {e}")
                    times.append(np.nan)
            
            results.append({
                "n_sites": n_sites,
                "solver": solver_name,
                "mean_time": np.nanmean(times),
                "std_time": np.nanstd(times),
                "min_time": np.nanmin(times),
                "max_time": np.nanmax(times),
            })
            
            print(f"    Mean time: {np.nanmean(times):.4f}s")
    
    return pd.DataFrame(results)


def run_gradient_speed_test(n_sites_list: List[int] = [10, 50, 100],
                            n_years: int = 1,
                            n_repeats: int = 3) -> pd.DataFrame:
    """
    Speed test differentiation through ODE solvers.
    This is critical for Bayesian inference with greta/TensorFlow.
    """
    
    results = []
    t_span = (0.0, 365.0 * n_years)
    
    for n_sites in n_sites_list:
        print(f"\nGradient test with N = {n_sites} sites...")
        
        # Create model with trainable parameters
        m_base = tf.Variable(np.random.lognormal(1.0, 0.5, n_sites).astype(np.float32))
        
        params = {
            "m_base": m_base,
            "amp1": tf.constant(0.6, dtype=tf.float32),
            "amp2": tf.constant(0.3, dtype=tf.float32),
            "a": tf.constant(0.3, dtype=tf.float32),
            "b": tf.constant(0.5, dtype=tf.float32),
            "c": tf.constant(0.8, dtype=tf.float32),
            "r": tf.constant(1/14, dtype=tf.float32),
            "g": tf.constant(1/10, dtype=tf.float32),
        }
        
        initial_state = tf.constant(
            [[0.01] * n_sites, [0.0] * n_sites],
            dtype=tf.float32
        )
        
        t_eval = tf.linspace(0.0, 365.0 * n_years, 100)
        
        # Test gradient computation with DormandPrince
        print("  Testing gradient with DormandPrince...")
        times = []
        
        for rep in range(n_repeats):
            try:
                start = time.perf_counter()
                
                with tf.GradientTape() as tape:
                    model = RossMacDonaldODE(n_sites, params)
                    solver = tfp.math.ode.DormandPrince(atol=1e-5, rtol=1e-5)
                    solution = solver.solve(model, t_span[0], initial_state,
                                           solution_times=t_eval)
                    # Loss: sum of final human infection proportions
                    loss = tf.reduce_sum(solution.states[-1, 0, :])
                
                grads = tape.gradient(loss, m_base)
                
                end = time.perf_counter()
                times.append(end - start)
                
            except Exception as e:
                print(f"    Error: {e}")
                times.append(np.nan)
        
        results.append({
            "n_sites": n_sites,
            "test_type": "gradient",
            "mean_time": np.nanmean(times),
            "std_time": np.nanstd(times),
        })
        
        print(f"    Mean gradient time: {np.nanmean(times):.4f}s")
    
    return pd.DataFrame(results)


# ============================================================================
# MAIN EXECUTION
# ============================================================================

if __name__ == "__main__":
    print("=" * 60)
    print("TensorFlow ODE Solver Speed Test")
    print("Ross-MacDonald Model with Time-Varying Parameters")
    print("=" * 60)
    
    # Check TensorFlow version
    print(f"\nTensorFlow version: {tf.__version__}")
    print(f"TensorFlow Probability version: {tfp.__version__}")
    
    # Run forward solve speed tests
    print("\n" + "=" * 60)
    print("FORWARD SOLVE SPEED TEST")
    print("=" * 60)
    forward_results = run_speed_test(
        n_sites_list=[10, 50, 100, 500],
        n_years=4,
        n_repeats=5
    )
    
    print("\n" + "-" * 40)
    print("Forward Solve Results Summary:")
    print("-" * 40)
    print(forward_results.to_string(index=False))
    
    # Run gradient speed tests
    print("\n" + "=" * 60)
    print("GRADIENT COMPUTATION SPEED TEST")
    print("=" * 60)
    gradient_results = run_gradient_speed_test(
        n_sites_list=[10, 50, 100],
        n_years=1,
        n_repeats=3
    )
    
    print("\n" + "-" * 40)
    print("Gradient Results Summary:")
    print("-" * 40)
    print(gradient_results.to_string(index=False))
    
    # Save results
    forward_results.to_csv("forward_solve_results.csv", index=False)
    gradient_results.to_csv("gradient_results.csv", index=False)
    print("\nResults saved to CSV files.")

